mbedtls/src/ssl/context.rs

```diff
@@ -127,21 +149,35 @@ impl Context {
             );
 
             self.io = Some(io);
+            self.inner.reset_handshake();            
+        }
+
+        self.handshake()
+    }
+}
 
-            self.handshake_cert.clear();
-            self.handshake_pk.clear();
-            self.handshake_ca_cert = None;
-            self.handshake_crl = None;
-            
-            match ssl_handshake(self.into()).into_result() {
-                Err(e) => {
-                    // safely end borrow of io
-                    ssl_set_bio(self.into(), ::core::ptr::null_mut(), None, None, None);
-                    self.io = None;
-                    Err(e)
-                },
-                Ok(_) => {
-                    Ok(())
+impl<T> Context<T> {
+    fn handshake(&mut self) -> Result<()> {
```

Comment:
Since we own `T` I think it's not really necessary to unset the IO in case of error anymore. This greatly simplifies the function.